exporters/failure/collector_pagerduty.py

```diff
@@ -0,0 +1,162 @@
+from pelorus.config.log import REDACT, log
+from pelorus.errors import FailureProviderAuthenticationError
+from pelorus.timeutil import parse_assuming_utc, second_precision
+from pelorus.utils import TokenAuth, set_up_requests_session
+
+_DATETIME_FORMAT = "%Y-%m-%dT%H:%M:%SZ"
+
+
+@define(kw_only=True)
+class PagerdutyFailureCollector(AbstractFailureCollector):
+    """
+    PagerDuty implementation of a FailureCollector
+    """
+
+    token: str = field(
+        default="",
+        metadata=env_vars(*env_var_names.TOKEN) | log(REDACT),
+        repr=False,
+    )
+
+    tls_verify: bool = field(default=True)
+
+    session: requests.Session = field(factory=requests.Session, init=False)
+
+    incident_urgency: set[str] = field(
+        factory=set,
+        converter=comma_or_whitespace_separated(set),
+        metadata=env_vars("PAGERDUTY_URGENCY"),
+    )
+
+    incident_priority: set[str] = field(
+        factory=set,
+        converter=comma_or_whitespace_separated(set),
+        metadata=env_vars("PAGERDUTY_PRIORITY"),
+    )
+
+    def __attrs_post_init__(self):
+        # disable .netrc
+        self.session.trust_env = False
+
+        if self.token:
+            set_up_requests_session(
+                self.session,
+                self.tls_verify,
+                auth=TokenAuth(self.token, is_pagerduty=True),
+            )
+
+    def get_incidents(self) -> list[dict]:
+        logging.info("Getting incidents")
+        url = "https://api.pagerduty.com/incidents?date_range=all&limit=10000"
+        headers = {
+            "Accept": "application/vnd.pagerduty+json;version=2",
+        }
+
+        resp = self.session.get(url, headers=headers)
+        try:
+            resp.raise_for_status()
+            # TODO too much noise?
+            logging.debug("PagerDuty successfully returned %s", resp.text)
+            return resp.json()["incidents"]
+        except requests.HTTPError as error:
+            if resp.status_code == requests.codes.unauthorized:
+                raise FailureProviderAuthenticationError from error
+            raise
+
+    def filter_by_urgency(self, urgency: str) -> bool:
+        if not self.incident_urgency:
+            return True
+        return urgency in self.incident_urgency
+
+    def filter_by_priority(self, priority: str) -> bool:
+        if not self.incident_priority:
+            return True
+        try:
+            return priority["summary"] in self.incident_priority
+        except TypeError:
+            return "null" in self.incident_priority
+
+    def search_issues(self) -> list[TrackerIssue]:
+        """
+        To maintain consistency, we are call this function `search_issues`,
+        but an `issue` in PagerDuty is called `incident`.
+        """
+        production_incidents = []
+        for incident in self.get_incidents():
+            is_production_bug = self.filter_by_urgency(
+                incident["urgency"]
```

Comment:
for any `incident["something"]`, we should check if the `something` exists in the incident, otherwise we will get KeyError, right?

Replace the incident["something"] with incident.get("something"), which returns None if it's not in the incident.

Or wrap this function around KeyError and ignore incident which has missing key and log it as warning.

I am concerned that some key may be missing with so many to be used in this loop.

Reply:
well it should not be missing, its the API spec. Do you really want to handle this?